test_app/tests/lib/redis/test_client.py

```diff
@@ -396,3 +398,89 @@ def test_redis_standalone_removes_cluster_settings():
         rm.assert_called_once
         assert 'host' in rm.call_args.kwargs
         assert 'cluster_error_retry_attempts' not in rm.call_args.kwargs
+
+
+@pytest.mark.parametrize(
+    "the_class",
+    [
+        DABRedis,
+        DABRedisCluster,
+    ],
+)
+def test_redis_clients_have_flushdb_mixin(the_class):
+    assert issubclass(the_class, DABRedisRespectACLFlushMixin)
+
+
+def test_flushdb_warns_about_async(expected_log):
+    args = {'mode': 'standalone'}
+    from ansible_base.lib.redis.client import RedisClientGetter
+
+    with mock.patch('redis.Redis.__init__', return_value=None):
+        client_getter = RedisClientGetter()
+        client = client_getter.get_client('rediss://localhost', **args)
+        client.keys = mock.MagicMock(return_value=[])
+        client.delete = mock.MagicMock(return_value=None)
+        with expected_log('ansible_base.lib.redis.client.logger', 'warning', "DABRedis clients implement an ACL friendly FLUSHDB which can not be async"):
+            client.flushdb(asynchronous=True)
+
+
+class notADABClient(DABRedisRespectACLFlushMixin, Redis):
+    pass
+
+
+def test_flushdb_not_a_dab_client():
+    with mock.patch('redis.Redis.__init__', return_value=None):
+        with pytest.raises(ImproperlyConfigured):
+            client = notADABClient('rediss://localhost')
+            client.flushdb()
+
+
+def test_ensure_flushdb_does_not_die_on_no_permission_exception():
+    args = {'mode': 'standalone'}
+    from redis.exceptions import NoPermissionError
+
+    from ansible_base.lib.redis.client import RedisClientGetter
+
+    with mock.patch('redis.Redis.__init__', return_value=None):
+        client_getter = RedisClientGetter()
+        client = client_getter.get_client('rediss://localhost', **args)
+        client.keys = mock.MagicMock(return_value=[])
```

Comment:
Keys needs to return something so that `delete` gets called.

Reply:
For this particular test I was more concerned with how the code handles the NoPermisionError but that being said the failing test indicates that there are no keys in Redis to delete so I think I need to wrap the delete in an `if` condition to ensure there is 1 or more keys to be deleted so this test will need to change.